adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -0,0 +1,379 @@
+            formalCheck.complete();
+        } else {
+            formalCheck.fail(new AmqpErrorException(AmqpError.DECODE_ERROR, "malformed payload"));
+        }
+        formalCheck.compose(ok -> {
+            switch (EndpointType.fromString(context.getEndpoint())) {
+            case TELEMETRY:
+                LOG.info(String.format("Received request to upload telemetry data to endpoint [{}]",
+                        context.getEndpoint()));
+                doUploadMessage(context, getTelemetrySender(context.getTenantId()),
+                        TelemetryConstants.TELEMETRY_ENDPOINT);
+                break;
+            case EVENT:
+                LOG.info(String.format("Received request to upload events at {}", context.getEndpoint()));
+                doUploadMessage(context, getEventSender(context.getTenantId()), EventConstants.EVENT_ENDPOINT);
+                break;
+            default:
+                return Future.failedFuture(new AmqpErrorException(AmqpError.NOT_ALLOWED, "unknown endpoint"));
+            }
+            return Future.succeededFuture();
+        }).otherwise(t -> {
+            AmqpErrorException error = (AmqpErrorException) t;
+            MessageHelper.rejected(context.delivery(), error.asErrorCondition());
+            return Future.failedFuture(t);
+        });
+    }
+
+    private void doUploadMessage(final AmqpContext context, final Future<MessageSender> senderFuture,
+            final String endpointName) {
+
+        final Future<JsonObject> tokenFuture = getRegistrationAssertion(context.getTenantId(), context.getDeviceId(),
+                null);
+        final Future<TenantObject> tenantConfigFuture = getTenantConfiguration(context.getTenantId());
+
+        CompositeFuture.all(tenantConfigFuture, tokenFuture, senderFuture).compose(ok -> {
+            TenantObject tenantObject = tenantConfigFuture.result();
+            if (tenantObject.isAdapterEnabled(context.getTenantId())) {
+                final MessageSender sender = senderFuture.result();
+                final Message downstreamMessage = newMessage(context.getResourceIdentifier(),
+                        sender.isRegistrationAssertionRequired(),
+                        endpointName, context.getMessageContentType(), context.getMessagePayload(),
+                        tokenFuture.result(), null);
+
+                if (context.delivery().remotelySettled()) {
+                    // client uses AT_MOST_ONCE delivery semantics -> fire and forget
+                    return sender.send(downstreamMessage);
+                } else {
+                    // client uses AT_LEAST_ONCE delivery semantics
+                    return sender.sendAndWaitForOutcome(downstreamMessage);
+                }
+            } else {
+                // this adapter is not enabled for tenant
+                return Future.failedFuture(
+                        new AmqpErrorException(AmqpError.NOT_ALLOWED, "This adapter is not enabled for tenant."));
```

Comment:
IMHO we should use a ClientErrorException with a 403 code here in order to be consistent with the other adapters' forwarding code.

Reply:
ok